python/src/iceberg/avro/decoder.py

```diff
@@ -0,0 +1,193 @@
+STRUCT_SIGNED_SHORT = struct.Struct(">h")  # big-endian signed short
+STRUCT_SIGNED_INT = struct.Struct(">i")  # big-endian signed int
+STRUCT_SIGNED_LONG = struct.Struct(">q")  # big-endian signed long
+
+
+class BinaryDecoder:
+    """Read leaf values."""
+
+    _input_stream: InputStream
+
+    def __init__(self, input_stream: InputStream) -> None:
+        """
+        reader is a Python object on which we can call read, seek, and tell.
+        """
+        self._input_stream = input_stream
+
+    def read(self, n: int) -> bytes:
+        """
+        Read n bytes.
+        """
+        if n < 0:
+            raise ValueError(f"Requested {n} bytes to read, expected positive integer.")
+        read_bytes = self._input_stream.read(n)
+        if len(read_bytes) != n:
+            raise ValueError(f"Read {len(read_bytes)} bytes, expected {n} bytes")
+        return read_bytes
+
+    def read_boolean(self) -> bool:
+        """
+        a boolean is written as a single byte
+        whose value is either 0 (false) or 1 (true).
+        """
+        return ord(self.read(1)) == 1
+
+    def read_int(self) -> int:
+        """int values are written using variable-length, zigzag coding."""
+        return self.read_long()
+
+    def read_long(self) -> int:
+        """long values are written using variable-length, zigzag coding."""
+        b = ord(self.read(1))
+        n = b & 0x7F
+        shift = 7
+        while (b & 0x80) != 0:
+            b = ord(self.read(1))
+            n |= (b & 0x7F) << shift
+            shift += 7
+        datum = (n >> 1) ^ -(n & 1)
+        return datum
+
+    def read_float(self) -> float:
+        """
+        A float is written as 4 bytes.
+        The float is converted into a 32-bit integer using a method equivalent to
+        Java's floatToIntBits and then encoded in little-endian format.
+        """
+        return float(STRUCT_FLOAT.unpack(self.read(4))[0])
+
+    def read_double(self) -> float:
+        """
+        A double is written as 8 bytes.
+        The double is converted into a 64-bit integer using a method equivalent to
+        Java's doubleToLongBits and then encoded in little-endian format.
+        """
+        return float(STRUCT_DOUBLE.unpack(self.read(8))[0])
+
+    def read_decimal_from_bytes(self, precision: int, scale: int) -> decimal.Decimal:
+        """
+        Decimal bytes are decoded as signed short, int or long depending on the
+        size of bytes.
+        """
+        size = self.read_long()
+        return self.read_decimal_from_fixed(precision, scale, size)
+
+    def read_decimal_from_fixed(self, precision: int, scale: int, size: int) -> decimal.Decimal:
+        """
+        Decimal is encoded as fixed. Fixed instances are encoded using the
+        number of bytes declared in the schema.
+        """
+        datum = self.read(size)
+        unscaled_datum = 0
+        msb = struct.unpack("!b", datum[0:1])[0]
+        leftmost_bit = (msb >> 7) & 1
```

Comment:
@Fokko, do you know why this uses `ord` and `unpack`? I don't think it needs to be this complicated. If you access a byte, then it is returned as the unsigned value.

```python
datum = b'\x01\x02\x03\xFF'
datum[0] # => 1
datum[-1] # => 255
```

Maybe I'm missing something, but I don't see how the line to produce `msb` isn't equivalent to `datum[0]`. I think this could be:

```java
    unscaled_datum = 0
    if datum[0] & 0x80 != 0:
        # negative two's complement
        ...
    else:
        # positive
        ...
```

Reply:
There may also be a much better way to do this, according to the answers on two's complement here: https://stackoverflow.com/questions/1604464/twos-complement-in-python

Reply:
LOL, you're right. I just copied this code. We can simplify it to:
```python
    def read_decimal_from_fixed(self, precision: int, scale: int, size: int) -> decimal.Decimal:
        """
        Decimal is encoded as fixed. Fixed instances are encoded using the
        number of bytes declared in the schema.
        """
        data = self.read(size)
        unscaled_datum = int.from_bytes(data, byteorder="big", signed=True)
        decimal_context.prec = precision
        return decimal_context.create_decimal(unscaled_datum).scaleb(
            -scale, decimal_context
        )
```
And that's what I did. Also added a couple of tests (should add more).

Reply:
I read the comment below, it is now:
```python
    def read_decimal_from_fixed(self, precision: int, scale: int, size: int) -> decimal.Decimal:
        """
        Decimal is encoded as fixed. Fixed instances are encoded using the
        number of bytes declared in the schema.
        """
        data = self.read(size)
        unscaled_datum = int.from_bytes(data, byteorder="big", signed=True)
        return unscaled_to_decimal(unscaled_datum, scale)
```

Reply:
Wow, quite a bit smaller!